jdaviz/core/data_formats.py

```diff
@@ -112,3 +122,145 @@ def identify_data(filename, current=None):
         raise ValueError('Mismatch between input file format and loaded configuration.')
 
     return valid_format, config
+
+
+def _get_wcs(filename, header):
+    """
+    Get gwcs.wcs.WCS or astropy.wcs.WCS from FITS file.
+    """
+    try:
+        with asdf_in_fits.open(filename) as af:
+            wcs = af.tree['meta']['wcs']
```

Comment:
I feel this. I thought about something like
```python
     with asdf_in_fits.open(filename) as af:
         meta = getattr(af.tree, 'meta', fits.getheader(filename))
```
but thought that was too obfuscated.

Reply:
I actually like that idea and since its a key and not an attribute, it might be _slightly_ less ugly (although on second thought you will need two layers of dictionaries which makes it a bit ugly again 😬 ). This entire `_get_wcs` function could be removed and

```
header = fits.getheader(filename, ext=ext)
data = fits.getdata(filename, ext=ext)
wcs = _get_wcs(filename, header)
```

replaced with:

```
data = fits.getdata(filename, ext=ext)
with asdf_in_fits.open(filename) as af:
    meta = af.tree.get('meta', {}).get('wcs', fits.getheader(filename, ext=ext)
```

which also avoids having to parse the header just to have on hand for the fallback.

Reply:
I agree @kecnry that this is a better way to do it, but even I have trouble reading it. 🤔